File: 202.happy-number.py
#
# @lc app=leetcode id=202 lang=python3
#
# [202] Happy Number
#

# @lc code=start
from functools import cache


class Solution:
    def isHappy(self, n: int) -> bool:
        if n < 10:
            return n == 1 or n == 7

        def digit_sum(num):
            res = 0
            while num:
                res += (num % 10) ** 2
                num //= 10
            return res
        
        while n != 1 and n != 4:
            n = digit_sum(n)
        return n == 1


# @lc code=end

File: 202.happy-number.rs
/*
 * @lc app=leetcode id=202 lang=rust
 *
 * [202] Happy Number
 */

// @lc code=start
use std::collections::HashSet;
impl Solution {
    pub fn is_happy(mut n: i32) -> bool {
        let digsum = move |n: i32| -> i32 {
            let mut s = 0;
            while n > 0 {
                s += (n % 10).pow(2);
                n /= 10;
            }
            s
        };
        loop {
            let mut s = 0;
            while n > 0 {
                s += (n % 10).pow(2);
                n /= 10;
            }
            match s {
                1 | 4 => break s == 1, // 4 results in a loop
                _ => n = s,
            }
        }
    }
}
// @lc code=end
